### PR TITLE
Cuckoofilter storage to disk and loading (WIP)

### PR DESCRIPTION
**Pull request by seiflotfy**
This is a work in progress:
- [x] store cuckoo filter buckets and fingerprints to disk
- [x] load cuckoo filter state from disk
- [x] cleanup

### CODE REVIEW

**Comment by martinpinto** [comment]
+1

File: counters/abstract/abstract.go [hunk 1]
```diff
@@ -16,9 +16,10 @@ type Counter interface {
 Info ...
 */
 type Info struct {
-	ID       string `json:"id"`
-	Type     string `json:"type"`
-	Capacity uint64 `json:"capacity"`
+	ID       string            `json:"id"`
+	Type     string            `json:"type"`
+	Capacity uint64            `json:"capacity"`
+	State    map[string]uint64 `json:"state"`
```
Comment (mbarkhau) [inline counters/abstract/abstract.go]: I'm still not sure about this. Some other domain types might want to have a different state. I think this is fine for now though, we can always change it later.

File: counters/manager_test.go [hunk 2]
```diff
@@ -211,6 +211,47 @@ func TestDumpLoadDefaultData(t *testing.T) {
 	}
 }
 
+func TestDumpLoadPurgableInfo(t *testing.T) {
```
Comment (mbarkhau) [inline counters/manager_test.go]: If we have a Purgeable counter, shouldn't we also be testing the Purge operation? Or is that not implemented yet?

File: counters/wrappers/cuckoofilter/cuckoofilter/bucket.go [hunk 3]
```diff
@@ -36,3 +38,46 @@ func (b *bucket) getFingerprintIndex(fp fingerprint) int {
 	}
 	return -1
 }
+
+type buckets struct {
+	id         string
+	numBuckets uint
+}
+
+func newBuckets(id string, numBuckets uint) (*buckets, error) {
```
Comment (mbarkhau) [inline counters/wrappers/cuckoofilter/cuckoofilter/bucket.go]: Ideally we would have more parameters, but I think you mentioned, that bucket and fingerprint sizes would be hardcoded for now.

File: counters/wrappers/cuckoofilter/cuckoofilter/cuckoofilter.go [hunk 4]
```diff
@@ -8,38 +12,40 @@ const maxCuckooCount = 500
...
 		capacity = 1
 	}
-	buckets := make([]bucket, capacity, capacity)
-	for i := range buckets {
-		buckets[i] = [bucketSize]fingerprint{}
-	}
-	return &CuckooFilter{buckets, 0}
+	//FIXME: return error
+	bs, _ := newBuckets(info.ID, capacity)
+	return &CuckooFilter{info.ID, bs, uint(info.State["count"])}
```
Comment (mbarkhau) [inline counters/wrappers/cuckoofilter/cuckoofilter/cuckoofilter.go]: Who is responsible for setting the initial "count" to zero?

Comment (seiflotfy) [inline counters/wrappers/cuckoofilter/cuckoofilter/cuckoofilter.go]: Can you rase these remakes as issues so we don’t forget to implement them?


—

File: counters/wrappers/cuckoofilter/domain.go [hunk 5]
```diff
@@ -12,29 +16,35 @@ var logger = utils.GetLogger()
...
+	lock sync.RWMutex
 }
 
 /*
 NewDomain ...
 */
-func NewDomain(info abstract.Info) Domain {
-	return Domain{info, cuckoofilter.NewCuckooFilter(uint(info.Capacity))}
+func NewDomain(info *abstract.Info) (*Domain, error) {
+	return &Domain{info, cuckoofilter.NewCuckooFilter(info), sync.RWMutex{}}, nil
```
Comment (mbarkhau) [inline counters/wrappers/cuckoofilter/domain.go]: Putting a Mutex on the domain seems like the wrong place for dealing with synchronization. Maybe we can have a goroutine which takes a channel where updates to domains are inserted and the goroutine takes care of the locking.

File: server/server_test.go [hunk 6]
```diff
@@ -44,7 +44,7 @@ func tearDownTests() {
 	storage.CloseInfoDB()
 }
 
-func http_request(s *Server, t *testing.T, method string, domain string, body string) *httptest.ResponseRecorder {
+func httpRequest(s *Server, t *testing.T, method string, domain string, body string) *httptest.ResponseRecorder {
```
Comment (mbarkhau) [inline server/server_test.go]: Ups, sorry about that. Old habits...

**Comment by mbarkhau** [comment]
Overall I don't think any of my objections should prevent this from being merged, so +1

I'll create some issues as you suggested.

File: counters/wrappers/cuckoofilter/domain.go (hunk 5 above)
Comment (seiflotfy) [inline counters/wrappers/cuckoofilter/domain.go]: Each request made via http is its own go routine. Golang default :D

**Pull request merged**